src/OrleansAzureUtils/Providers/Streams/AzureQueue/AzureQueueAdapterReceiver.cs

```diff
@@ -79,10 +82,38 @@ public Task Initialize(TimeSpan timeout)
                 outstandingTask = task;
                 IEnumerable<CloudQueueMessage> messages = await task;
 
-                List<IBatchContainer> azureQueueMessages = messages
-                    .Select(msg => (IBatchContainer)AzureQueueBatchContainer.FromCloudQueueMessage(msg, lastReadMessage++)).ToList();
+                if (this.hasLargeMessageSupport)
+                {
+                    var list = new List<IBatchContainer>();
+                    foreach (var segment in messages.Select(m => MessageSegment.FromByteArray(m.AsBytes)))
+                    {
+                        List<MessageSegment> segmentList;
+                        if (segment.Count == 1)
+                        {
+                            list.Add(AzureQueueBatchContainer.FromMessageSegments(new[] {segment}, lastReadMessage++));
+                            continue;
+                        }
+
+                        if (!messageSegments.TryGetValue(segment.Guid, out segmentList))
+                        {
+                            segmentList = new List<MessageSegment>();
+                            messageSegments.Add(segment.Guid, segmentList);
+                        }
+
+                        segmentList.Add(segment);
+                        if (segment.Count == segmentList.Count)
```

Comment:
how do we know we're not getting duplicate segments?  Azure queue will redeliver messages that have not been removed from the queue, so duplicates can show up, correct?